pkg/logger/logger.go

```diff
@@ -1,57 +1,92 @@
 package logger
 
 import (
+	"errors"
 	"flag"
+	"fmt"
 	"os"
+	"strconv"
+	"strings"
+	"sync/atomic"
 
 	"github.com/go-logr/logr"
+	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
-	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	ctrlzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
 )
 
-// NewNamedLogger creates a new logger for a component.
-// If the mode is set (so can be different from the default one),
-// it will create a new logger with the specified mode's options.
-func NewNamedLogger(log logr.Logger, name string, mode string) logr.Logger {
-	if mode != "" {
-		log = NewLogger(mode)
-	}
-	return log.WithName(name)
+var logLevel atomic.Value
+
+// copy from controller-runtime/pkg/log/zap/flag.go.
+var levelStrings = map[string]zapcore.Level{
+	"debug": zap.DebugLevel,
+	"info":  zap.InfoLevel,
+	"error": zap.ErrorLevel,
 }
 
-func NewLoggerWithOptions(mode string, override *zap.Options) logr.Logger {
-	opts := newOptions(mode)
-	overrideOptions(opts, override)
-	return newLogger(opts)
+// adjusted copy from controller-runtime/pkg/log/zap/flag.go, keep the same argument name.
+func stringToLevel(flagValue string) (zapcore.Level, error) {
+	level, validLevel := levelStrings[strings.ToLower(flagValue)]
+	if validLevel {
+		return level, nil
+	}
+	logLevel, err := strconv.Atoi(flagValue)
+	if err != nil {
+		return 0, fmt.Errorf("invalid log level \"%s\"", flagValue)
+	}
+	if logLevel > 0 {
+		intLevel := -1 * logLevel
+		return zapcore.Level(int8(intLevel)), nil
+	}
+
+	return 0, fmt.Errorf("invalid log level \"%s\"", flagValue)
 }
 
-// in DSC component, to use different mode for logging, e.g. development, production
-// when not set mode it falls to "default" which is used by startup main.go.
-func NewLogger(mode string) logr.Logger {
-	return newLogger(newOptions(mode))
+func SetLevel(levelStr string) error {
+	if levelStr == "" {
+		return nil
+	}
+	levelNum, err := stringToLevel(levelStr)
+	if err != nil {
+		return err
+	}
+
+	// ctrlzap.addDefauls() uses a pointer to the AtomicLevel,
+	// but ctrlzap.(*levelFlag).Set() the structure itsef.
+	// So use the structure and always set the value in newOptions() to addDefaults() call
+	level, ok := logLevel.Load().(zap.AtomicLevel)
+	if !ok {
+		return errors.New("stored loglevel is not of type *zap.AtomicLevel")
+	}
+
+	level.SetLevel(levelNum)
+	return nil
 }
 
-func newLogger(opts *zap.Options) logr.Logger {
-	return zap.New(zap.UseFlagOptions(opts))
+func NewLogger(mode string, override *ctrlzap.Options) logr.Logger {
+	opts := newOptions(mode)
+	overrideOptions(opts, override)
+	logLevel.Store(opts.Level)
```

Comment:
so the override call does get all: level stack output etc, but the Store only set the level?
so maybe the function name is misleading, even it is only called in "main" with flag option passed in.

Reply:
yes. overrideOptions overrides all possible zap options which are supposed to be changed by command line (controller framework functionality) to make them priority preconfigured by logmode values. Then the logger is created with the resulting options.

But dynamically only loglevel will be changed from DSCI, so it's stored for later amendments. It's an interface, consists of 2 parts, so atomic.Value used to save it.